Update evaluation metrics for sample forecasts

Fuhan-Yang opened the pull request:
- Updated iup.eval.py: 
1. Cleaned up previous metrics
2. Added a new function `abs_diff()` that returns a function to calculate absolute difference between predictions and observed data at specific date. 
3. Added a new function `summarize_score()` to loop through all score functions given observed data and predictions. Thanks for insights from @afmagee42 !

- Updated scripts/eval.py:

1. Converted prediction distribution into quantiles, allowing them to compare with observed data. 
2.  Get the functions that are returned by `abs_diff()`, and calculate together with `mspe()`. 

A long format of pl.DataFrame is returned, with `score_name` and `score_value`, grouped by `model`, `forecast_start` and `quantile`. Note that now model needs to be specified as `HillModel` to avoid the error (`lium` returns negative estimates), but this is fine as we will only use logistic + linear model. 
 
- Updated config.yaml: 
1. To enable users to select the date to calculate the difference for `abs_diff()`.
2. To specify which quantile from the prediction distribution to calculate the metrics

- Updated Makefile
- Updated test functions

Resolves #154

eschrom commented on iup/eval.py:
```diff
@@ -48,51 +47,83 @@ def score(
+    all_scores = pl.DataFrame()
+    for score_name in score_funs:
+        score = joined_df.select(
+            quantile=pl.col("quantile").unique().first(),
```
Looking at the eval.py script, maybe predictions should never be passed into `summarize_score` with more than one unique quantile? If so, should that be checked with an `assert`? And could we just use `.first()` without `.unique()`?

Fuhan-Yang replied:
Your interpretation is exactly right, and the suggestion sounds good! Yes, `summarize_score` function only takes `QuantileForecast`, this means `SampleForecast` has to summarize to quantile beforehand. And only the `QuantileForecast` with a unique quantile should enter `summarize_score()`. So it is a good idea to assert these conditions and drop `unique()` in the function.

Fuhan-Yang replied:
Now I'm looking at this, `summarize_score()` requires to be strictly embedded in a for loop through quantiles, maybe this is not the most ideal structure, but we can use this for now.

eschrom commented on scripts/eval.py:
```diff
@@ -7,47 +7,80 @@
-                    "The forecast and the test data do not have the same number of dates."
+                score_funcs = {}
+
+                if config["scores"]["difference_by_date"] is not None:
```
This is super educational!

Fuhan-Yang replied:
I'm glad it helps! Thanks for inputs from @afmagee42 too!

eschrom commented on scripts/eval.py:
```diff
@@ -7,47 +7,80 @@
-                    score_fun=pl.lit(score_name),
+                scores = eval.summarize_score(test, summary_pred, score_funcs)
+
+                scores = scores.with_columns(
```
I notice that the model name is added as column at the script level, but the forecast date is added as a column inside iup/eval.py. This means that the "forecast_start" column is the earliest date in the test data, not necessarily exactly the forecast date the user requested (which is adjusted to match the test data). 

I don't know if that's a problem here, but it's at least worth thinking about: do we want the "forecast_start" column to be exactly what the user requested (which is what we use to write file names for diagnostics, etc.) or instead what the actual first date in the test data is?

Fuhan-Yang replied:
I prefer to have the actual first date in the test data, because the user should know that the forecast and the evaluation is done at not exactly (but close to) the date they choose. We can note this in the final product, saying due to the interval of reporting period, the data is not always available on the date the users request. I also think the scope of this discrepancy is small, because the user only specifies the start and end date of forecast, instead of specifying all the dates within the forecast period. And hopefully this is not common, at least for the selected period, the actual first date of the test data matches the forecast start date in config.

Fuhan-Yang replied:
This is sensible to me! So the user knows that the package runs as they expect, but the data on that date is just realistically not available. I like that!

eschrom commented on scripts/postprocess.py:
```diff
@@ -83,7 +78,7 @@ def plot_individual_projections(
     )
 
-    return alt.layer(pred_chart, obs_chart, data=plot_pred).facet(
+    return alt.layer(pred_chart, obs_chart, data=pred).facet(
```
As mentioned above, we also must be able to facet over grouping factors within a model (which would produce multiple pages of faceted plots, I suppose).

Fuhan-Yang replied:
Yes, this will be my next step, refactoring the postprocess.py given the new code in evaluation. More to come!

eschrom commented on scripts/postprocess.py:
```diff
@@ -120,20 +115,14 @@ def plot_summary(
-    )
+    plot_pred = pred.with_columns(
+        lower=pl.col("estimate")
+        .quantile(lci / 100)
```
I thought uptake should always be on 0-1 scale by this point (so we don't divide by 100), but maybe I'm wrong - just something to check!

Fuhan-Yang replied:
This is a good point, I will fix this in the next PR in refactoring postprocess.py!

eschrom reviewed (changes requested):
There are two main threads to my comments:
1) Accommodating grouping factors within models (e.g. age group, state)
2) Using cumulative, not incident, data and forecasts for evaluation.

Plus several miscellaneous other things. If you want to talk through any of the requested changes, let me know!

Also, the function generator for calculating absolute difference on different dates is super cool!

Fuhan-Yang commented on iup/eval.py:
```diff
@@ -35,10 +36,8 @@ def check_date_match(data: IncidentUptakeData, pred: PointForecast):
-    pred: PointForecast,
-    score_fun: Callable[[pl.Expr, pl.Expr], pl.Expr],
+def summarize_score(
+    data: IncidentUptakeData, pred: QuantileForecast, score_funs: Dict[str, Callable]
```
Thanks for the detailed explanation! Sorry it is me who forgot to change to CumulativeUptakeData. I focused on changing the code without noticing that I adapted previous type and type hinthing. It turned out either IncidentUptakeData and CumulativeUptakeData can run without error, so I missed this. Thanks for catching!

Fuhan-Yang commented on iup/eval.py:
```diff
@@ -48,51 +47,83 @@ def score(
-        data.join(pred, on="time_end", how="inner", validate="1:1")
-        .rename({"estimate": "data", "estimate_right": "pred"})
-        .select(
+    joined_df = data.join(pred, on="time_end", how="inner", validate="1:1").rename(
```
Good catch! I plan to add a for loop on grouping factors. Then the final score will also be grouped by the grouping factors, in addition to forecast_start, quantile, and model name. The grouping factors should be the same as `config["data"]["groups"]`, because we should evaluate the models exactly as how they are trained. What do you think?

eschrom replied:
> The grouping factors should be the same as `config["data"]["groups"]`

Exactly! Totally agree.

Fuhan-Yang commented on iup/eval.py:
```diff
@@ -16,15 +17,15 @@ def check_date_match(data: IncidentUptakeData, pred: PointForecast):
     # sort data and pred by date #
     data = IncidentUptakeData(data.sort("time_end"))
-    pred = PointForecast(pred.sort("time_end"))
+    pred = QuantileForecast(pred.sort("time_end"))
```
That's a good point! Already added groups argument in `check_date_match`

Fuhan-Yang commented on iup/eval.py:
```diff
@@ -48,51 +47,83 @@ def score(
+    )
+
+    all_scores = pl.DataFrame()
+    for score_name in score_funs:
```
It will work! `for score_name in score_funs` is essentially the same as `for score_name in score_funs.keys()`

Fuhan-Yang commented on iup/eval.py:
```diff
@@ -48,51 +47,83 @@ def score(
-    )
+            score_name=pl.lit(score_name),
+            score_value=score_funs[score_name](pl.col("data"), pl.col("pred")),
+        ).filter(pl.col("score_value").is_not_null())
```
The function returned by `abs_diff()` is a polar expression that returns `None` when it's not the selected date (see line 127 in iup.eval.py). When embedded in a polars data frame, it adds a new column where absolute difference only shows in the selected date, while other dates in `time_end` is `None`. 

It's not a red flag, it's just a little awkward in polars when it's required to return a polar expression.

Fuhan-Yang commented on scripts/eval.py:
```diff
@@ -7,47 +7,80 @@
+        config file to specify the expected quantile from the sample distribution and evaluation metrics to calculate
+
+    Returns:
+        A pl.DataFrame with score name and score values, grouped by model, forecast start and quantile
```
Already done!

Fuhan-Yang commented on scripts/eval.py:
```diff
@@ -7,47 +7,80 @@
+    Arguments:
+    data:
+        observed data with at least "time_end" and "estimate" columns
+    pred:
```
Good catch! Yes otherwise it will fail

Fuhan-Yang commented:
Updates: 
- Corrected data type to `CumulativeUptakeData`
- Enabled evaluation function to calculate metrics at group level
- Updated relevant tests.
- Other small tweaks

Thanks Ed for thorough review! Please let me know if you have further suggestions!

Fuhan-Yang commented on scripts/eval.py:
```diff
@@ -4,50 +4,93 @@
+
+
+def eval_all_forecasts(
+    data: pl.DataFrame, pred: SampleForecast, config: dict
```
`SampleForecast` is imported by itself in line 7, so no need to call iup here!

Fuhan-Yang commented on scripts/eval.py:
```diff
@@ -4,50 +4,93 @@
+                        for date in config["scores"]["difference_by_date"]
+                    }
+
+                score_funcs[config["scores"]["others"]] = getattr(
```
Your understanding is correct! And good catch! Already added the option to loop through multiple metrics in `config["scores"]["others"]`, and check if this is `None`.
